### ISSUE TITLE
What is the main benefit over the JBoss tool?

### ISSUE THREAD

**Issue by keilw** [kevto/tck-coverage#1]
JBoss created https://github.com/jboss/jboss-test-audit some while ago. And beside its own JSRs (CDI, Bean Validation) at least [JSR 354](https://github.com/JavaMoney/jsr354-tck) ([JavaMoney](http://javamoney.org)) and [363/385](https://github.com/unitsofmeasurement/unit-tck) ([Units of Measurement](http://unitsofmeasurement.github.io/)) make use of it, too. It was last updated in May 2017, so it looks well maintained for such a utility.

**Comment by kevto** [comment]
Hi. A few months back Ed mentioned to Frank on suggesting JUnit 5 for our TCK. As I looked through the TCKs you mentioned, to figure out on a good start, I saw most of the TCKs were created using TestNG. I was so under the impression that JBoss' tool required TestNG (should have looked better, my bad) in order to run, I thought it would be better to create something similar which leverages the JUnit 5 ecosystem.

At where I work, we've got a large codebase that contains over thousands and thousands unit tests created using JUnit 3 and 4. We leverage the ecosystem to create reports similarly to what this project does but only with JUnit 5 and the ability to make use of more writers using the same interface and is able to do in much more details. As of now, I've only implemented the text file writer.

The major differences between JBoss' tool and mine is that mine is very depended on JUnit 5. I did keep modularization in my mind and shouldnt take much effort to leverage TestNG's ecosystem instead I think. 

- Mine also is much easier to embed into your project and requires a single properties file. JBoss' would require hardcoded configuration and additional classes to run the TCK;
- JBoss' allows the usage of the Test annotation from TestNG rather than a custom annotation like mine does but I'm making the assumption that TckTest annotation is simple to understand.

Regardless of the execution order of the unit tests (which shouldn't matter anyhow), the output will be ordered properly. In the near future (within months) it will have support to be ran in any maven and gradle project that implements the specification, by leveraging the build tools' API.

So in short:
- Easier to embed into your project;
- Multiple output writers;
- Leverages JUnit 5 to allow you to only write the unit tests rather than worrying by the integration of the TCK coverage report generator.

**Comment by keilw** [comment]
I don't see how the JBoss tool would require JUnit or TestNG. It just happened to be used by CDI and JavaMoney. And for JSR 363 (to some extent CDI also does that when testing aspects only relevant in the Full JEE Profile as opposed to the Web Profile) what was even more important is the ease of  grouping tests and combining those groups into profiles.
`@SpecAssertion(section = "4.5", id = "45-A1")` in the JBoss annotation is not significantly different from this one. The extra XML file may seem like an overhead, but it adds long free text descriptions that can be maintained there without touching the code, so I would not see that as a problem.

Much of the JBoss tool was contributed by Spec Leads or former Spec Leads like @pmuir or @gunnarmorling, so maybe you could discuss improvements with them and consider a newer version of their tool ;-)

**Comment by gunnarmorling** [comment]
Hey, just wanted to point out one big plus of the tooling we developed: the XML file used as input for the TCK coverage report is automatically generated from the spec document itself. By marking TCK-relevant phrases in the spec doc (via AsciiDoc roles), they will be added automatically to that file, greatly helping with keeping the spec and the TCK in sync. Most recently, the MVC 1.0 EG has taken up that tooling AFAIK.

I think the usage of JUnit vs. TestNG is mostly a technicality, shouldn't be hard to generalize our tooling (if it's even specific) in that regard. All in all, I'd recommend a collaboration on a joint project instead of creating a new one from scratch.

**Comment by kevto** [comment]
Hi all, TestNG's annotation is mentioned in plain text somewhere in the code of the tool as I just found out but the impact is very minimal (groups?). I think it could be easily extended with an equivalent for JUnit.

From my understanding now is that the coverage report of JBoss' tool does not include the test results of the unit tests as I thought it'd. Correct me if I'm wrong but the coverage report simply validates if there are any methods using the SpecAssertion annotation with the proper section id? Technically the spec document is the leading factor and validates if we're not missing any unit tests that doesn't cover the spec right? 

I'll implement our TCK using JBoss' and contribute to the tool rather than maintaining my tool on my own instead.

**Comment by gunnarmorling** [comment]
> Correct me if I'm wrong but the coverage report simply validates if there are any methods using the SpecAssertion annotation with the proper section id?

Yes, exactly. Reporting of actual test failures is left to the used testing tool (TestNG).

> I'll implement our TCK using JBoss' and contribute to the tool rather than maintaining my tool on my own instead.

That sounds great.

Let me also summon @gsmet as he's been the one who worked on and with the TCK tooling most recently for the Bean Validation 2.0 spec. E.g. he had been working on some scripts for updating section ids in the tests after spec updates. Also @chkal and friends may have done some additions in the context of the MVC spec.

**Comment by chkal** [comment]
> Also @chkal and friends may have done some additions in the context of the MVC spec.

That's correct. Our workflow is basically like this:
  * The assertion XML file [is generated](https://github.com/mvc-spec/mvc-spec/blob/433a1344d64c2afd5fbbdc00f1acc31b5f713ec1/spec/pom.xml#L126-L163) from the specification document and published together with the other artifacts to the Sonatype OSS snapshot repository
  * We use a [custom XSTL file](https://github.com/mvc-spec/mvc-spec/blob/master/spec/src/main/xsl/tck-audit.xsl) which allows us to explicitly set the assertion IDs which are referenced via `@SpecAssertion` in the TCK
  * The TCK [gets the latest assertion XML](https://github.com/mvc-spec/mvc-tck/blob/c050db27be170ff8fc7f0f41ab61e45c6b69fcc7/tests/pom.xml#L112-L138), [generates the report](https://github.com/mvc-spec/mvc-tck/blob/c050db27be170ff8fc7f0f41ab61e45c6b69fcc7/tests/pom.xml#L112-L138) and automatically [uploads the report](https://github.com/mvc-spec/mvc-tck/blob/master/.travis/deploy-coverage-report.sh) to the spec homepage
  * We are currently using JUnit 4 for the TCK, mostly because AFAIK Arquillian doesn't support JUnit 5 yet.

**Comment by kevto** [comment]
That's nice, thanks for your comment. I will certainly take a deep dive into your TCK once we reach the stage of implementing our TCK.

**Comment by chkal** [comment]
You're welcome! Feel free to contact me if you have any questions.

**Comment by keilw** [comment]
Glad to get all of you in touch. Especially with Jakarta EE and a whole trove of Open Source TCKs it seems good to discuss what could be improved.
